File: 543-diameter-of-binary-tree/diameter-of-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int diameterOfBinaryTree(TreeNode* root) {
        int maxD=0;

         dfs(root,maxD);
         return maxD;

        
    }

    int dfs(TreeNode* root, int& maxD)
    {
        if(!root) return 0;
        int left=dfs(root->left, maxD);
        int right=dfs(root->right, maxD);

        maxD=max(maxD, left+right);

        return 1+max(left,right);
    }
};

File: 543-diameter-of-binary-tree/diameter-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:

        # hya madhe tar left nd right madhe jaa and check kar heigh kuthe jyast aahe
        # i cannot do top down because i do not know which side to go for maximizing
        maxD=[0]*1
        maxD[0]=0
        def dfs(node):
            #mala hyat each child varu varti kahi tari send karaycha aahe
            if not node:
                return 0
            left=dfs(node.left)
            right=dfs(node.right)

            maxD[0]=max(left+right,maxD[0])
            
            return 1+max(left,right)


        dfs(root)
        return maxD[0]
        